app/assets/javascripts/reducers/training.js

```diff
@@ -100,6 +103,11 @@ export default function training(state = initialState, action) {
         enabledSlides: [...state.enabledSlides, data.slide.id],
         completed: data.completed
       };
+    case SET_TRAINING_MODE:
+      return {
+        ...state,
+        editMode: JSON.parse(mainDiv.getAttribute('data-training_mode')).editMode
```

Comment:
This seems like violation of the Redux state model, as it changes state based on things outside of the current state and the data included in the action object (so it's not a pure function). Although it wouldn't actual be mutable normally, I think it still ought to be refactored so that reading from the DOM happens outside of the reducer and the editMode value is included in the action.

Reply:
Thank you for catching that Redux violation. I've refactored the code to maintain reducer purity by moving the DOM interaction out of the reducer. The editMode value is now read in the action creator and passed through the action's payload. The reducer now only depends on the state and action parameters, following Redux's pure function principles.